debian/contrib/shutdown.8
.TH SHUTDOWN 8 "August 29, 2021" "" "GNU/Linux System Adminstrator's manual"
.SH NAME
shutdown, reboot, poweroff \- poweroff or reboot the system
.SH SYNOPSIS
.B /sbin/shutdown [-r] [-f] [-F]
.br
.B /sbin/reboot [-w] [-f] [-n]
.br
.B /sbin/halt [-w] [-f] [-n]
.br
.B /sbin/poweroff [-w] [-f] [-n]
.SH DESCRIPTION
.BR Shutdown
is a program to poweroff or reboot the system that maintains some compatibility with
original SysV-init halt, poweroff, reboot and shutdown programs.
These programs are expected by some initscripts, graphical desktop environments and tools like acpi.
.RE
When called as shutdown, halt or poweroff without options,
.BR runit(8)
is told to shutdown the system and poweroff.
.RE
When called as reboot
.BR runit(8)
is told to reboot the system.
.RE
When
.BR runit(8)
is not the current init system this program sends data in the appropriate format to perform the requested action to the initctl pipe, if it exists.
.SH SHUTDOWN OPTIONS

.TP
.B \-r
Reboot the system instead of poweroff.

.TP
.B \-f
Write a
.B /fastboot
flag file. The program or the scripts responsible for the system boot task can test for this file when the system comes up again and decide to skip fsck.
The program or the scripts that perform the boot task also need to take care of the removal of the flag file.

.TP
.B \-F
Write a
.B /forcefsck
flag file. The program or the scripts responsible for the system boot task can test for this file when the system comes up again and decide to run fsck with a 'force' flag.
The program or the scripts that perform the boot task also need to take care of the removal of the flag file.

.SH HALT OPTIONS

.TP
.B \-f
Invoke sync(), then force an unsafe reboot or poweroff immediately without signaling the init system.
This will likely result in an unclean shutdown and can cause data loss or corruption.
When runit is init, this option is a No-Op and all other options are ignored. See #899246

.TP
.B \-\-force
Invoke sync(), then force an unsafe reboot or poweroff immediately without signaling the init system.
This will likely result in an unclean shutdown and can cause data loss or corruption.
This option works regardless of the running init system.

.TP
.B \-w, \-\-wtmp\-only
No-Op, maintained for compatibility with initscripts. See #919699

.TP
.B \-n
Write a
.B /etc/runit/nosync
flag file. If this file exists runit does not invoke sync() before reboot or poweroff.
The
.B /etc/runit/nosync
flag file is also checked when this program is called with
.B \-f
or
.B \-\-force

.SH SWITCHING FORM OTHER INIT SYSTEMS
This program maintains a compatibility layer with SysV-init's initctl pipe according to the spec described in SysV-init's initctl(5). This allow one to reboot the system when switching from another init to runit-init.
.RE
Currently only switching from systemd and SysV-init is tested but any other init system that maintains an initctl pipe compatible with SysV's one should work.

.SH BUGS
Combining flags, like
.B halt -wf
is not supported, all merged short options will be ignored.
.RE
Non existent or unsupported options are silently ignored.

.SH SEE ALSO
.BR init (8)


debian/changelog
runit (2.1.2-43) unstable; urgency=medium

  * default-syslog: add a comment in the run file to
     clarify the intended usage.
  * Move the nosync file back to /etc/runit/:
     it was moved to /run by mistake (Closes: #993602)
     + thanks to András Korn
  * Stop using sysv-rc in stage 3: 'run_sysv_scripts' can
     be easily used for shutdown tasks
  * Adjust runit-init Depends:
     - drop sysv-rc: it's still impossible to remove it because
       of initscripts
     - add insserv, it's needed to get the right boot sequence
       for initscripts and sysvinit scripts
     - bump the minimal required runit version to 2.1.2-43,
       because of changes in 'run_sysv_scripts' and stage 3
  * Update runit-init and runit-systemd description
  * source only changes:
     - move all contrib services under debian/sv
     - stop patching upstream stage{1,2}, we already
       maintain our version of stage{1,2,3} in d/contrib
     - rename quilt patches to match the order in which
       they are applied

 -- Lorenzo Puliti <plorenzo@disroot.org>  Tue, 07 Sep 2021 13:26:47 +0200

runit (2.1.2-42) unstable; urgency=medium

  * Release to unstable
  * getty-run: add hvc0 service, disabled by default:
      this is usually needed by Xen hypervisor
  * Add a default-syslog virtual service:
      to increase portability, services that log to syslog
      can depend on this rather than on a specific
      syslog daemon
  * shutdown.c:
      - try to fix FTBFS on Hurd (Closes: #992629)
      - distinguish between halt and shutdown flags:
        The -r flag ( as reboot) now only works with shutdown;
        Add a -f flag to shutdown (skip fsck next boot);
        Add -F flag to shutdown (force fsck next boot);
        The -f (as force a shutdown) and -w/--wtmp-only flags now
         only work with halt.
        (Closes: #992631)
      - Add a -n flag to halt, to skip sync() before reboot/poweroff;
        also, always check for runit.nosync flag file before invoking
        sync()  (Closes: #992641)
      - make halt '-f' flag a noop when runit is init. This way runit
        can use its own code to reboot/poweroff the system; users
        can still use the long '--force' option to force a shutdown
        without signaling the init (Closes: #899246)
  * Update shutdown(8) manpage
  * Bump Standards Version to 4.6.0, no change required
  * Update lintian overrides for 'alternative init but not init.d script'

 -- Lorenzo Puliti <plorenzo@disroot.org>  Sun, 29 Aug 2021 13:54:45 +0200

runit (2.1.2-41exp1) experimental; urgency=medium

  * Set rules-requires-root to no
  * Runit-run: drop dependency on runit-systemd | sysvinit-core
  * Runit: get rid of ancient transition code in prerm
  * Add upstream metadata
  * Add support for policy-rc.d hack in invoke-run
  * Update invoke-run manpage for policy-rc.d
  * Update News files
  * Shutdown.c:
    - fix wrong command line parsing logic that always
      caused a No-Op when any option was given, breaking,
      among other things, the init switch from Sysvinit
      (Closes: #991227)
    - reboot the system with -r flag (Closes: #990774)
  * Update shutdown(8) manpage
  * Update license and copyright years

 -- Lorenzo Puliti <plorenzo@disroot.org>  Fri, 16 Jul 2021 20:35:34 +0200

runit (2.1.2-40) unstable; urgency=medium

  * Ack previous NMU, thanks Boyuan Yang for the upload
  * Getty-run: default to agetty, only suggests fgetty (Closes: #981248)
  * invoke-run: don't stop sysv script if is a symlink
    + many thanks to Devuan testers

 -- Lorenzo Puliti <plorenzo@disroot.org>  Tue, 16 Feb 2021 10:38:41 +0100

runit (2.1.2-39.1) unstable; urgency=medium

  * Non-maintainer upload.
  * No change source-only upload to allow migration to testing.

 -- Boyuan Yang <byang@debian.org>  Fri, 25 Dec 2020 16:31:31 -0500

runit (2.1.2-39) unstable; urgency=medium

  * Enable shutdown on ctrl-alt-del keyboard request
  * Bump Standards Version to 4.5.1
  * Reintroduce transitional runit-systemd package (Closes: #976187)
  * Add a news file for runit-init

 -- Lorenzo Puliti <plorenzo@disroot.org>  Sun, 20 Dec 2020 11:16:49 +0100

runit (2.1.2-38) unstable; urgency=medium

  * Clean up at the end of the forced-rescan-test
  * Fix some minor lintian tag
  * Bump dh compat level to 13
  * d/rules: fix broken autopkgtest phony target
  * d/runit.NEWS: document recent changes
  * Do not restart the gettys on upgrade
  * Suggests socklog with runit
  * Release to unstable

 -- Lorenzo Puliti <plorenzo@disroot.org>  Wed, 25 Nov 2020 03:05:05 +0100

runit (2.1.2-37) experimental; urgency=medium

  * Bring back the test about forced-rescan
  * Fix forced rescan test failure on sbuild (Closes: #941322)
  * Clean the 'current' symlink on purge
  * Invoke-run: add a DEBUG option to finish-default
  * Finish-default: print invoke-run instead of runsv
  * Get non default rundirectory from kernel command line
  * Add a 'solo' rundirectory for containers (Closes: #950851)
  * Merge runit-sysv and runit-systemd into runit-run (Closes: #953875)
  * Add lintian overrides for manpage without executable
  * Update copyright for debian files
  * Add an autopkgtest to catch regression on #953875
  * runit-run: don't stop the service on upgrade (systemd)
  * runit-run: rescan inittab on install/remove (sysvinit)

 -- Lorenzo Puliti <plorenzo@disroot.org>  Thu, 10 Sep 2020 14:35:26 +0200

runit (2.1.2-36) unstable; urgency=medium

  * Use dot-symlink to mark a service as disabled (Closes: #942320)
  * Minor improvements to invoke-run (Closes: #943395)
  * Add finish-default (Closes: #943397)
  * Update invoke-run manpage for finish-default
  * Bump Standards Version to 4.5.0:
    - Create '_runit-log' user
  * New maintainer
  * Add a lintian override for runit

 -- Lorenzo Puliti <plorenzo@disroot.org>  Thu, 27 Feb 2020 12:12:07 +0100

runit (2.1.2-35) unstable; urgency=medium

  * Temporary disable forced-rescan test (Closes: #941273)


debian/contrib/shutdown.c
// Copyright 2018 Dmitry Bogatov <KAction@gnu.org>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

// This source file compiles to following binaries:
//
//  - /sbin/shutdown
//  - /sbin/halt
//  - /sbin/reboot

// They must work when /sbin/init is runit, but PID1 is not. It happens
// when bin:runit-init replaces bin:systemd-sysv or bin:sysvinit-core.
//
// Process of system rebooting/halting has nothing to do with PID1, it
// is all about reboot(2).
//
// reboot(8)/halt(8), provided by sysvinit, when invoked without -f,
// ask PID1 to terminate all processes. As last step, PID1 invokes
// /etc/rc0.d/K08halt or /etc/rc6.d/K08reboot, which, at their turn,
// invoke reboot(8)/halt(8) with -f option. With that option on,
// reboot(2) is called, causing Linux to reboot/halt.
//
// They are not needed for work in command line, since 'runit' uses
// another, more simple interface to system
// but they have been here

// /sbin/reboot binaries. Strictly speaking, they are not
#include <stdbool.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <sys/reboot.h>
#include <assert.h>
#include <string.h>
#include <signal.h>

#define STOPIT "/run/runit.stopit"
#define NOSYNC "/etc/runit/nosync"
#define DOFSCK "/forcefsck"
#define NOFSCK "/fastboot"

#define write2(msg) write(2, msg, sizeof(msg))
typedef enum shutdown_action { ACTION_REBOOT, ACTION_HALT } shutdown_action;

/* Copied and slightly simplified from sysvinit-2.88dsf:/src/initreq.h */
struct sysv_request {
	int	magic;			/* Magic number                 */
	int	cmd;			/* What kind of request         */
	int	runlevel;		/* Runlevel to change to        */
	int	sleeptime;		/* Time between TERM and KILL   */
	char	unused[368];
};
#define SYSV_MAGIC              0x03091969
#define SYSV_CMD_RUNLVL         1
#define SYSV_FIFO               "/run/initctl"

/* Since we are forced to support more and more of command line options
 * expected by initscripts, let us make it more structured.
 */
struct config {
	enum shutdown_action action;
	bool force;
	bool nosync;
	bool nofsck;
	bool dofsck;
	bool noop;		/*stop and exit 0 without actually reboot or poweroff  */
};

struct stat s;

/* Luckily, systemd supports sysvinit pipe interface, so there
 * is no need to write separate function interfacing systemd.
 */
static bool
sysv_shutdown(shutdown_action action)
{
	struct sysv_request request;
	request.magic = SYSV_MAGIC;
	request.cmd = SYSV_CMD_RUNLVL;

	switch (action) {
	case ACTION_REBOOT: request.runlevel = '6'; break;
	case ACTION_HALT:   request.runlevel = '0'; break;
	default: return false;
	};

	int fd = open(SYSV_FIFO, O_WRONLY);
	if (fd < 0) {
		return false;
	}
	bool ok = write(fd, &request, sizeof(request)) == sizeof(request);
	close(fd);
	return ok;
}

static bool
syscall_shutdown(shutdown_action action)
{
	if (stat(NOSYNC,&s) == -1) {
		sync();
	}
	switch (action) {
	case ACTION_REBOOT:
		reboot(RB_AUTOBOOT);
		break;
	case ACTION_HALT:
#ifdef RB_POWER_OFF
		reboot(RB_POWER_OFF);
#else
#ifdef RB_HALT_SYSTEM
		reboot(RB_HALT_SYSTEM);
#else
#ifdef RB_HALT
		reboot(RB_HALT);
#else
		reboot(RB_AUTOBOOT);
#endif
#endif
#endif
		break;
	}
	return false;
}

static bool
runit_shutdown(shutdown_action action)
{
	char* args[] = {
		"/sbin/init", (action == ACTION_HALT) ? "0" : "6", NULL };
	execv("/sbin/init", args);
	return false;
}

int open_trunc(const char *fn)
{ return open(fn,O_WRONLY | O_NDELAY | O_TRUNC | O_CREAT,0644); }

static void
parse_command_line(struct config *cfg, int argc, char **argv)
{
	int i;

	cfg->action    = ACTION_HALT;
	cfg->force     = false;
	cfg->nosync = false;
	cfg->dofsck = false;
	cfg->nofsck = false;
	cfg->noop = false;

	if (strstr(argv[0], "reboot")) {
		cfg->action = ACTION_REBOOT;
	}

	for (i = 1; i != argc; ++i) {
		if (strstr(argv[0], "shutdown")) {
			if (strcmp(argv[i], "-r") == 0) {
				cfg->action = ACTION_REBOOT;
				continue;
			}
			if (strcmp(argv[i], "-f") == 0) {
				cfg->nofsck = true;
				continue;
			}
			if (strcmp(argv[i], "-F") == 0) {
				cfg->dofsck = true;
				continue;
			}
		}
		else {
			if (strcmp(argv[i], "-f") == 0) {
				if (stat(STOPIT,&s) == -1) {
					cfg->force = true;
				}
				else {
					cfg->noop = true;
					write2("WARNING: -f is a noop when runit is init\n");
				}
				continue;
			}
			if (strcmp(argv[i], "--force") == 0) {
				cfg->force = true;
				continue;
			}
			if (strcmp(argv[i], "-w") == 0) {
				write2("WARNING: writing wtmp with -w is not supported for now\n");
				cfg->noop = true;
				continue;
			}
			if (strcmp(argv[i], "--wtmp-only") == 0) {
				cfg->noop = true;
				continue;
			}
			if (strcmp(argv[i], "-n") == 0) {
				cfg->nosync = true;
				continue;
			}
		}
	}
}

int
main(int argc, char **argv)
{
	struct config cfg;
	bool ok;

	parse_command_line(&cfg, argc, argv);

	if (geteuid() != 0) {
		write2("shutdown: must be superuser.\n");
		return 2;
	}
	if (cfg.noop) {
		// No-Op: see #919699 and #899246
		return 0;
	}
	if (cfg.dofsck) {
		if (close(open_trunc(DOFSCK) == -1)) {
			write2("Failed to write forcefsck file.\n");
			return 1;
		}
	}
	if (cfg.nofsck) {
		if (close(open_trunc(NOFSCK) == -1)) {
			write2("Failed to write fastboot file.\n");
			return 1;
		}
	}
	if (cfg.nosync) {
		if (close(open_trunc(NOSYNC) == -1)) {
			write2("Failed to write runit.nosync file.\n");
			return 1;
		}
	}
	if (cfg.force) {
		ok = syscall_shutdown(cfg.action);
	} else {
		ok = sysv_shutdown(cfg.action)
			|| runit_shutdown(cfg.action);
		sleep(15);
		/* System should be already down, but still. */
		syscall_shutdown(cfg.action);
	}
	return ok ? 0 : 1;
}
